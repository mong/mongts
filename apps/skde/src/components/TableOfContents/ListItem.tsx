import React from "react";
import { useIntersectionByID } from "../../helpers/hooks/useintersectionobserver";
import style from "./ListItem.module.css";

type ListItemProps = {
  children?: React.ReactNode;
  expanded?: string;
  setExpanded?: React.Dispatch<React.SetStateAction<string>>;
  href: string;
  linkTitle: string;
  i?: string;
};

export const ListItem: React.FC<ListItemProps> = ({
  children,
  href,
  linkTitle,
  expanded,
  setExpanded,
}) => {
  const isVisbile = useIntersectionByID(href.replace("#", ""), "0px");

  return (
    <li
      className={`${style.tocListItem} ${isVisbile ? style.active : isVisbile}`}
      data-testid="tocItem"
    >
      <a
        href={href}
        onClick={() => {
          setExpanded((state) => (state === href ? "none" : href));
        }}
        data-testid={`toc_a_${href}`}
      >
        {linkTitle}
      </a>
      {(expanded === href || expanded === "all") && children}
    </li>
  );
};
