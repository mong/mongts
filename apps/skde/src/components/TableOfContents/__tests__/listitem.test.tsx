import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { ListItem } from "../ListItem";
import { vi, describe, it, expect } from "vitest";
import "@testing-library/jest-dom";

describe("ListItem component", () => {
  it("renders list item with valid props", () => {
    const props = {
      href: "#test",
      linkTitle: "Test Link",
      expanded: "",
      setExpanded: vi.fn(),
    };
    const { getByTestId } = render(<ListItem {...props} />);
    expect(getByTestId("tocItem")).toBeInTheDocument();
    expect(getByTestId("toc_a_#test")).toHaveAttribute("href", "#test");
  });

  it("renders list item with children", () => {
    const props = {
      href: "#test",
      linkTitle: "Test Link",
      expanded: "all",
      setExpanded: vi.fn(),
      children: <div>Test Children</div>,
    };
    const { getByTestId, getByText } = render(<ListItem {...props} />);
    expect(getByTestId("tocItem")).toBeInTheDocument();
    expect(getByTestId("toc_a_#test")).toHaveAttribute("href", "#test");
    expect(getByText("Test Children")).toBeInTheDocument();
  });

  it("renders list item with expanded state", () => {
    const props = {
      href: "#test",
      linkTitle: "Test Link",
      expanded: "#test",
      setExpanded: vi.fn(),
    };
    const { container } = render(<ListItem {...props} />);
    expect(container).toMatchSnapshot();
  });

  it("click on link toggles expanded state", () => {
    const props = {
      href: "#test",
      linkTitle: "Test Link",
      expanded: "",
      setExpanded: vi.fn(),
    };
    const { getByTestId } = render(<ListItem {...props} />);
    const link = getByTestId(`toc_a_#test`);
    fireEvent.click(link);
    expect(props.setExpanded).toHaveBeenCalledTimes(1);
  });

  it('click on link with expanded state set to "all" does not toggle expanded state', () => {
    const props = {
      href: "#test",
      linkTitle: "Test Link",
      expanded: "all",
      setExpanded: vi.fn(),
    };
    const { getByTestId } = render(<ListItem {...props} />);
    const link = getByTestId(`toc_a_#test`);
    expect(props.setExpanded).not.toHaveBeenCalled();
    fireEvent.click(link);
    expect(props.setExpanded).toHaveBeenCalled();
  });
});
