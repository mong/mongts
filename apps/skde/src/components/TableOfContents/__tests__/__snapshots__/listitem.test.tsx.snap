// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListItem component > renders list item with expanded state 1`] = `
<div>
  <li
    class="_tocListItem_3a2e40 false"
    data-testid="tocItem"
  >
    <a
      data-testid="toc_a_#test"
      href="#test"
    >
      Test Link
    </a>
  </li>
</div>
`;
